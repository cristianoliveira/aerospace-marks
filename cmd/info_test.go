package cmd

import (
	"errors"
	"fmt"
	"testing"

	"github.com/cristianoliveira/aerospace-marks/internal/mocks"
	"github.com/cristianoliveira/aerospace-marks/internal/storage"
	"github.com/cristianoliveira/aerospace-marks/internal/testutils"
	"github.com/gkampitakis/go-snaps/snaps"
	"go.uber.org/mock/gomock"
)

func TestInfoCmd(t *testing.T) {
	t.Run("Happy path - all compatible", func(tt *testing.T) {
		ctrl := gomock.NewController(tt)
		defer ctrl.Finish()

		aerospaceConnection, aerospaceClient := mocks.MockAerospaceConnection(ctrl)
		dbClient, storageClient := mocks.MockStorageDbClient(ctrl)

		storageClient.
			EXPECT().
			Client().
			Return(dbClient).
			Times(1)

		dbClient.
			EXPECT().
			GetStorageConfig().
			Return(storage.StorageConfig{
				DbPath: "/tmp/database/",
				DbName: "foo.db",
			}).
			Times(1)

		gomock.InOrder(
			aerospaceConnection.
				EXPECT().
				GetSocketPath().
				Return("/tmp/foo.sock", nil).
				Times(1),

			aerospaceConnection.
				EXPECT().
				CheckServerVersion().
				Return(nil).
				Times(1),

			aerospaceConnection.
				EXPECT().
				GetServerVersion().
				Return("aerospace-ipc v0.1.0", nil).
				Times(1),
		)

		cmd := InfoCmd(
			storageClient,
			aerospaceClient,
		)
		out, err := testutils.CmdExecute(cmd)
		if err != nil {
			tt.Fatal(err)
		}

		cmdExecuted := fmt.Sprintf("aerospace-marks %s", cmd.Use)
		snaps.MatchSnapshot(tt, cmdExecuted, out)
	})

	t.Run("Happy path - non compatible", func(tt *testing.T) {
		ctrl := gomock.NewController(tt)
		defer ctrl.Finish()

		aerospaceConnection, aerospaceClient := mocks.MockAerospaceConnection(ctrl)
		dbClient, storageClient := mocks.MockStorageDbClient(ctrl)

		storageClient.
			EXPECT().
			Client().
			Return(dbClient).
			Times(1)

		dbClient.
			EXPECT().
			GetStorageConfig().
			Return(storage.StorageConfig{
				DbPath: "/tmp/database/",
				DbName: "foo.db",
			}).
			Times(1)

		gomock.InOrder(
			aerospaceConnection.
				EXPECT().
				GetSocketPath().
				Return("/tmp/foo.sock", nil).
				Times(1),

			aerospaceConnection.
				EXPECT().
				CheckServerVersion().
				Return(errors.New("incompatible version because reasons")).
				Times(1),

			aerospaceConnection.
				EXPECT().
				GetServerVersion().
				Return("aerospace-ipc v3.1.0", nil).
				Times(1),
		)

		cmd := InfoCmd(
			storageClient,
			aerospaceClient,
		)
		out, err := testutils.CmdExecute(cmd)
		if err != nil {
			tt.Fatal(err)
		}

		cmdExecuted := fmt.Sprintf("aerospace-marks %s", cmd.Use)
		snaps.MatchSnapshot(tt, cmdExecuted, out)
	})

	t.Run("Failure - to retrieve socket path", func(tt *testing.T) {
		ctrl := gomock.NewController(tt)
		defer ctrl.Finish()

		aerospaceConnection, aerospaceClient := mocks.MockAerospaceConnection(ctrl)
		dbClient, storageClient := mocks.MockStorageDbClient(ctrl)

		storageClient.
			EXPECT().
			Client().
			Return(dbClient).
			Times(1)

		dbClient.
			EXPECT().
			GetStorageConfig().
			Return(storage.StorageConfig{
				DbPath: "/tmp/database/",
				DbName: "foo.db",
			}).
			Times(1)

		gomock.InOrder(
			aerospaceConnection.
				EXPECT().
				GetSocketPath().
				Return("", errors.New("missing socket path")).
				Times(1),
		)

		cmd := InfoCmd(
			storageClient,
			aerospaceClient,
		)
		out, err := testutils.CmdExecute(cmd)
		if err == nil {
			tt.Fatal(err)
		}

		cmdExecuted := fmt.Sprintf("aerospace-marks %s", cmd.Use)
		snaps.MatchSnapshot(tt, cmdExecuted, "Output", out, "Error", err.Error())
	})
}
