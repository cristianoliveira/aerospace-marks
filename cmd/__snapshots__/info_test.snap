
[TestInfoCmd/Happy_path_-_all_compatible - 1]
aerospace-marks info
Aerospace Marks CLI - Configuration

[Socket]
Path: /tmp/foo.sock
Version: aerospace-ipc v0.1.0
Status: Compatible.

[Database]
Name: foo.db
Path: /tmp/database/

[Logging]
Path: /tmp/aerospace-marks.log
Level: DISABLED

Configure with ENV variables:
AEROSPACESOCK - Path to the socket file.
AEROSPACE_MARKS_DB_PATH - Path to database directory.
AEROSPACE_MARKS_LOGS_LEVEL - Log level [debug|info|warn|error] (default: disabled)
AEROSPACE_MARKS_LOGS_PATH - Path to the logs file.

---

[TestInfoCmd/Happy_path_-_non_compatible - 1]
aerospace-marks info
Aerospace Marks CLI - Configuration

[Socket]
Path: /tmp/foo.sock
Version: aerospace-ipc v3.1.0
Status: Incompatible. Reason: incompatible version because reasons

[Database]
Name: foo.db
Path: /tmp/database/

[Logging]
Path: /tmp/aerospace-marks.log
Level: DISABLED

Configure with ENV variables:
AEROSPACESOCK - Path to the socket file.
AEROSPACE_MARKS_DB_PATH - Path to database directory.
AEROSPACE_MARKS_LOGS_LEVEL - Log level [debug|info|warn|error] (default: disabled)
AEROSPACE_MARKS_LOGS_PATH - Path to the logs file.

---

[TestInfoCmd/Failure_-_to_retrieve_socket_path - 1]
aerospace-marks info
Output

Error
failed to get socket path: missing socket path
---
